### PR TITLE
Ensure fsencode/fsdecode works on ASCII FS.  Use tarball for unicode test files

### PR DESCRIPTION
**Pull request by pombredanne**
For #688 this should solved the paths encoding issue even on Doker images without locale and an ASCII FS encoding 

For #755 this ensures the repo can be cloned on Windows by using a tarball for test paths with weird encodings

Both as reported by @sschuberth : Thanks!

### CODE REVIEW

File: src/commoncode/fileutils.py [hunk 1]
```diff
@@ -77,7 +77,9 @@ def logger_debug(*args):
         return logger.debug(' '.join(isinstance(a, basestring) and a or repr(a) for a in args))
 
 
-FS_ENCODING = sys.getfilesystemencoding() or sys.getdefaultencoding()
+FS_ENCODING = sys.getfilesystemencoding()
+# normalize the encoding name
+FS_ENCODING = codecs.lookup(FS_ENCODING).name
```
Comment (sschuberth) [inline src/commoncode/fileutils.py]: Is this still the correct thing to do? I though when using a version of `backports.os` that has https://github.com/pjdelport/backports.os/pull/8 merged, we don't need to do normalization anymore?

File: thirdparty/prod/backports.os.ABOUT [hunk 2]
```diff
@@ -15,3 +15,6 @@ copyright: Copyright (c) Piët Delport, and Python Software Foundation
 notes: the invalid_utf8_indexes() helper is copied from Bob Ippolito's 
  pyutf8 package (pyutf8/ref.py) and is MIT-licensed
  https://github.com/etrepum/pyutf8/blob/master/pyutf8/ref.py
+ This verion is an advanced patched version fixing the #688 issue
+ and contain the code in the contributed pull request at 
+ https://github.com/pjdelport/backports.os/pull/7
```
Comment (sschuberth) [inline thirdparty/prod/backports.os.ABOUT]: Ah, I see. This custom version of `backports.os` still has your #7 merged, whereas we should now use a version that has #8 merged. I guess we can wait until #8 is merged upstream and then use a version straight from upstream.

**Comment by sschuberth** [comment]
The Unicode fixture `tests/scancode/data/unicode_fixtures/empty.txt` is present in the TAR but has not been removed from Git, and the fixture `tests/scancode/data/unicode_fixtures/snow ☃/.gitkeep` is *not* present in the TAR (only a `snow` directory is) and it has also not been removed from Git.

File: thirdparty/prod/backports.os.ABOUT (hunk 2 above)
Comment (pombredanne) [reply thirdparty/prod/backports.os.ABOUT]: @sschuberth yes, I will update this with the latest https://github.com/pjdelport/backports.os/pull/8 that @pjdelport kindly published to Pypi too.

**Comment by pombredanne** [comment]
@sschuberth you wrote:

> The Unicode fixture tests/scancode/data/unicode_fixtures/empty.txt is present in the TAR but has not been removed from Git, and the fixture tests/scancode/data/unicode_fixtures/snow ☃/.gitkeep is not present in the TAR (only a snow directory is) and it has also not been removed from Git.

Good catch. Will fix

**Comment by pombredanne** [comment]
@sschuberth  If I check https://github.com/nexB/scancode-toolkit/tree/688-fsencode-fsdecode-755-unicode-test-files/tests/scancode/data `unicode_fixtures/` is gone-gone as a dir?

And the archive seems to contains properly the needed files?
```
$ tar -tvf unicode_fixtures.tar.gz 
-rw-rw-r-- pombreda/pombreda 9 2017-09-17 21:18 unicode_fixtures/pets/names.txt
-rw-rw-r-- pombreda/pombreda 11 2017-09-17 21:18 unicode_fixtures/.hidden
-rw-rw-r-- pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/empty.txt
-rw-rw-r-- pombreda/pombreda  9 2017-09-17 21:18 unicode_fixtures/nums
-rw-rw-r-- pombreda/pombreda  3 2017-09-17 21:18 unicode_fixtures/foo bar
drwxrwxr-x pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/pets/
-rw-rw-r-- pombreda/pombreda 23 2017-09-17 21:18 unicode_fixtures/users/index.html
drwxrwxr-x pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/
-rw-rw-r-- pombreda/pombreda 18 2017-09-17 21:18 unicode_fixtures/todo.html
-rw-rw-r-- pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/snow ☃/.gitkeep
-rw-rw-r-- pombreda/pombreda  6 2017-09-17 21:18 unicode_fixtures/users/tobi.txt
drwxrwxr-x pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/users/
-rw-rw-r-- pombreda/pombreda 11 2017-09-17 21:18 unicode_fixtures/todo.txt
drwxrwxr-x pombreda/pombreda  0 2017-09-17 21:18 unicode_fixtures/snow ☃/
```

**Comment by pombredanne** [comment]
This branch now has the latest @pjdelport backports.os 1.1 !

File: src/commoncode/fileutils.py (hunk 1 above)
Comment (sschuberth) [reply src/commoncode/fileutils.py]: I see this change is still present with backports.os 0.1.1 in use. Shouldn't your either revert this change, or use the proper fsencode/fsdecode that you mentioned?

**Comment by sschuberth** [comment]
But is `thirdparty/prod/backports.os-0.1.patch1-py2-none-any.whl` still needed?

**Comment by sschuberth** [comment]
When downloading https://github.com/nexB/scancode-toolkit/raw/0ae20e45a735b660c60e5c1b2b8db449d83fc261/tests/scancode/data/unicode_fixtures.tar.gz and opening it with WinRAR, I still only see a `snow` directory:

snow

On Linux, it shows correctly, however, and also if I run `tar -tvf` from Git Bash. Very odd, I'll report this as a bug in WinRAR.

**Comment by sschuberth** [comment]
> I'll report this as a bug in WinRAR.

FYI, the RAR devs already confirmed the bug in WinRAR 5.5 realted to Unicode chars in TAR files with PAX extended headers. It will be fixed in the next version.

**Comment by pombredanne** [comment]
@sschuberth good catch on Winrar ;)
I do not use it though.
This latest commit should be all set now.

**Comment by sschuberth** [comment]
The code looks good to me now, but Windows CI still fails with

    WindowsError: [Error 123] The filename, directory name, or volume label syntax is incorrect: 'c:\\users\\appveyor\\appdata\\local\\temp\\1\\scancode_appveyor\\tst\\ rbabmi\\dqxinc\\unicode_fixtures.tar.gz\\unicode_fixtures\\snow ?'

**Comment by pombredanne** [comment]
All should be passing now: I am skipping some tests on Windows. Merging!

**Comment by pombredanne** [comment]
@sschuberth you wrote

> But I doubt that's true. I can download and extract https://raw.githubusercontent.com/nexB/scancode-toolkit/d5f3116f0ef8b21c81e60a0a7c18eb46bb257630/tests/scancode/data/unicode_fixtures.tar.gz on Windows just fine. It's probably just a matter of doing it right programmatically.

I can download and extract things alright on Windows eventually using 7zip or extractcode... but they will mangle the names such that they are valid windows names which defeats the purpose of the test: Windows cannot git clone this nor could I find a way to recreate some of the express test files either.
So I have tried many ways but there is no way some tests run on Windows.... so I am going to merge this as it is now.

**Pull request merged**